### PR TITLE
[Proposal] component model loading

### PR DESCRIPTION
**Pull request by dannypsnl**
#2943 loading

misc

1. involves `loadPreamble`
2. involves `loadVec<SectionType>(vector, F)`
3. involves `Expect<std::optional<T>> loadOption(std::function<Expect<void>(T &)> F)` to load syntax like `t?:<valtype>?`

### CODE REVIEW

**Comment by juntao** [comment]
Hello, I am a [code review bot](https://github.com/flows-network/github-pr-summary/) on [flows.network](https://flows.network/).

It could take a few minutes for me to analyze this PR. Relax, grab a cup of coffee and check back later. Thanks!

File: lib/loader/ast/component.cpp [hunk 1]
```diff
@@ -68,18 +75,129 @@ Expect<std::variant<AST::Component, AST::Module>> Loader::loadUnit() {
...
+      NestedComp->getVersion() = {Ver[0], Ver[1]};
+      NestedComp->getLayer() = {Ver[2], Ver[3]};
+      if (auto Res = loadCompnent(*NestedComp); !Res) {
+        return Unexpect(Res);
+      }
+      std::vector<std::unique_ptr<AST::Component>> &V =
+          Comp.getComponentSection().getContent();
+      V.push_back(std::move(NestedComp));
+      // FIXME: The compiler believe it copies the NestedComp, I can't get why
+      // for now.
```
Comment (dannypsnl) [inline lib/loader/ast/component.cpp:156]: Seems the trouble is starting from `include/ast/section.h`, not here

File: include/ast/section.h [hunk 2]
```diff
@@ -283,5 +283,16 @@ class AOTSection {
   /// @}
 };
 
+class Component;
+
+class ComponentSection : public Section {
+public:
+  /// Getter of content.
+  std::vector<Component *> &getContent() noexcept { return Content; }
+
```
Comment (dannypsnl) [inline include/ast/section.h:292]: The fix cannot be seemed as correct

1. `Span` is not provided here.
2. Now we have to ensure `Component *` didn't get misused.

File: include/loader/loader.h [hunk 3]
```diff
@@ -247,6 +253,11 @@ class Loader {
   enum class InputType : uint8_t { WASM, UniversalWASM, SharedLibrary };
   InputType WASMType = InputType::WASM;
   /// @}
+
+  // Metadata
+  std::vector<Byte> ModuleVersion = {0x01, 0x00, 0x00, 0x00};
+  // spec says 0x0a, but it's actually 0x0d, where cargo component compiled out
+  std::vector<Byte> ComponentVersion = {0x0d, 0x00, 0x01, 0x00};
```
Comment (dannypsnl) [inline include/loader/loader.h:260]: This will with a final change when component model is "complete"

**Comment by dannypsnl** [comment]
`IWYU checker / macOS` failed since it failed to install python 3.11....

**Comment by hydai** [comment]
> `IWYU checker / macOS` failed since it failed to install python 3.11....

Should we remove all possible `2to3*`?

**Comment by dannypsnl** [comment]
That's triggered by llvm17, can we control that?

**Comment by hydai** [comment]
> That's triggered by llvm17, can we control that?

Nope. So I guess we have to remove the 2-to-3, which is installed by other pre-shipped packages to make the llvm install correctly.

File: include/ast/component/sort.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+public:
+  Sort getSort() const noexcept { return S; }
+  Sort &getSort() noexcept { return S; }
+  uint32_t getSortIdx() const noexcept { return Idx; }
+  uint32_t &getSortIdx() noexcept { return Idx; }
+
+private:
+  Sort S;
+  uint32_t Idx;
+};
```
Comment (dannypsnl) [inline include/ast/component/sort.h:49]: reuse same class for SortIndex and CoreSortIndex

The problem here is in alias section, we want to load both; but in instance section, distinction can help us form better type, this is place that we might need higher type.

File: include/ast/module.h [hunk 5]
```diff
@@ -161,5 +138,55 @@ class Module {
...
 
+class CoreModuleSection : public Section {
+public:
+  /// Getter of content.
+  Span<const Module> getContent() const noexcept { return Content; }
+  std::vector<Module> &getContent() noexcept { return Content; }
+
+private:
+  std::vector<Module> Content;
+};
```
Comment (dannypsnl) [inline include/ast/module.h:149]: Here, unlike nested component, we no need to wrap module again (no recursive definition)

File: include/ast/component/sort.h (hunk 4 above)
Comment (dannypsnl) [reply include/ast/component/sort.h:49]: A potential solution here is using `std::variant`

File: lib/loader/filemgr.cpp [hunk 6]
```diff
@@ -70,6 +70,19 @@ Expect<Byte> FileMgr::readByte() {
...
+    return Unexpect(R);
+  }
+  if (*R != V) {
+    Pos--;
+    LastPos = Pos;
+    return false;
+  }
+  return true;
+}
+
```
Comment (dannypsnl) [inline lib/loader/filemgr.cpp:85]: note, better just remove it

File: lib/loader/ast/type.cpp [hunk 7]
```diff
@@ -69,14 +69,11 @@ Expect<void> Loader::loadType(AST::FunctionType &FuncType) {
...
-      return logLoadError(ErrCode::Value::IntegerTooLong, FMgr.getLastOffset(),
-                          ASTNodeAttr::Type_Function);
-    }
-  } else {
+  if (auto Res = FMgr.readByte(0x60U); !Res) {
     return logLoadError(Res.error(), FMgr.getLastOffset(),
                         ASTNodeAttr::Type_Function);
+  } else if (!(*Res)) {
+    return Unexpect(ErrCode::Value::IntegerTooLong);
   }
```
Comment (dannypsnl) [inline lib/loader/ast/type.cpp:77]: recover this

File: lib/loader/ast/component.cpp [hunk 8]
```diff
@@ -68,18 +75,142 @@ Expect<std::variant<AST::Component, AST::Module>> Loader::loadUnit() {
...
+      break;
+    case 0x01: {
+      auto ResPreamble = Loader::loadPreamble();
+      if (!ResPreamble) {
+        spdlog::error(ErrInfo::InfoAST(ASTNodeAttr::Component));
+        return Unexpect(ResPreamble);
+      }
+      auto WasmMagic = ResPreamble->first;
+      auto Ver = ResPreamble->second;
+      if (Ver != ModuleVersion) {
```
Comment (q82419) [inline lib/loader/ast/component.cpp:118]: unikely()

File: lib/loader/ast/component.cpp [hunk 9]
```diff
@@ -68,18 +75,142 @@ Expect<std::variant<AST::Component, AST::Module>> Loader::loadUnit() {
...
+      }
+      break;
+    case 0x04: {
+      auto ResPreamble = Loader::loadPreamble();
+      if (!ResPreamble) {
+        return Unexpect(ResPreamble);
+      }
+      auto WasmMagic = ResPreamble->first;
+      auto Ver = ResPreamble->second;
+      if (Ver != ComponentVersion) {
```
Comment (q82419) [inline lib/loader/ast/component.cpp:151]: unlikely

File: lib/loader/ast/comptype.cpp [hunk 10]
```diff
@@ -0,0 +1,650 @@
...
+    return logLoadError(ErrCode::Value::MalformedRecordType,
+                        FMgr.getLastOffset(), ASTNodeAttr::DefType);
+  }
+  auto RName = FMgr.readName();
+  if (!RName) {
+    return logLoadError(ErrCode::Value::MalformedRecordType,
+                        FMgr.getLastOffset(), ASTNodeAttr::DefType);
+  }
+  Label = *RName;
+  if (Label.size() != *RLen) {
```
Comment (q82419) [inline lib/loader/ast/comptype.cpp:29]: unlikely

File: lib/loader/ast/comptype.cpp [hunk 11]
```diff
@@ -0,0 +1,650 @@
...
+  }
+}
+
+Expect<void> Loader::loadType(List &Ty) { return loadType(Ty.getValType()); }
+
+Expect<void> Loader::loadType(Tuple &Ty) {
+  if (auto Res = loadVec<CompTypeSection>(
+          Ty.getTypes(),
+          [this](ValueType T) -> Expect<void> { return loadType(T); })) {
+    if (Ty.getTypes().size() == 0) {
```
Comment (q82419) [inline lib/loader/ast/comptype.cpp:143]: unlikely

File: include/ast/component/type.h [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+#include "ast/type.h"
+
+#include <optional>
+#include <vector>
+
+namespace WasmEdge {
+namespace AST {
+
+enum class PrimValType {
+  Bool,
```
Comment (q82419) [inline include/ast/component/type.h:28]: Assign the value number into this enum class (maybe uint8_t?)

File: lib/loader/ast/comptype.cpp [hunk 13]
```diff
@@ -0,0 +1,650 @@
...
+    Ty.getIndex() = *Res;
+  } else {
+    return Unexpect(Res);
+  }
+  return {};
+}
+
+Expect<void> Loader::loadType(Byte Tag, PrimValType &Ty) {
+  switch (Tag) {
+  case 0x7f: // bool
```
Comment (q82419) [inline lib/loader/ast/comptype.cpp:216]: We can use the value of  member of enum class to determine the malformed input.

File: lib/loader/ast/comptype.cpp [hunk 14]
```diff
@@ -0,0 +1,650 @@
...
+
+Expect<void> Loader::loadImportDecl(ImportDecl &Decl) {
+  if (auto Res = loadImportExportName(Decl.getImportName()); !Res) {
+    return Unexpect(Res);
+  }
+  return loadExternDesc(Decl.getExternDesc());
+}
+
+Expect<void> Loader::loadType(ResultList &Ty) {
+  if (auto RTag = FMgr.readByte(); !RTag) {
```
Comment (q82419) [inline lib/loader/ast/comptype.cpp:407]: if (auto RTag =  .... ) {
} else {
  return Unexpect(RTag);
}

File: lib/loader/ast/comptype.cpp [hunk 15]
```diff
@@ -0,0 +1,650 @@
...
+
+  return {};
+}
+
+Expect<void> Loader::loadType(LabelValType &Ty) {
+  // labelvaltype ::= l:<label'> t:<valtype>
+  if (auto Res = loadLabel(Ty.getLabel()); !Res) {
+    return Unexpect(Res);
+  }
+  if (auto Res = loadType(Ty.getValType()); !Res) {
```
Comment (q82419) [inline lib/loader/ast/comptype.cpp:78]: return loadType(...)

File: lib/loader/ast/comptype.cpp [hunk 16]
```diff
@@ -0,0 +1,650 @@
...
+Expect<void> Loader::loadType(Record &RecTy) {
+  // syntax:
+  //     lt*:vec(<labelvaltype>)
+  //
+  // output: (record (field lt)*)  (if |lt*|>0)
+  auto Res = loadVec<CompTypeSection>(
+      RecTy.getLabelTypes(),
+      [this](LabelValType LT) -> Expect<void> { return loadType(LT); });
+
+  if (Res) {
```
Comment (q82419) [inline lib/loader/ast/comptype.cpp:93]: if (auto Res = ...)

File: lib/loader/ast/comptype.cpp [hunk 17]
```diff
@@ -0,0 +1,650 @@
...
+  } else {
+    return Unexpect(Res);
+  }
+}
+Expect<void> Loader::loadType(Flags &Ty) {
+  if (auto Res = loadVec<CompTypeSection>(
+          Ty.getLabels(), [this](std::string Label) -> Expect<void> {
+            return loadLabel(Label);
+          })) {
+    if (Ty.getLabels().size() == 0) {
```
Comment (q82419) [inline lib/loader/ast/comptype.cpp:157]: unlikely

File: lib/loader/ast/instance.cpp [hunk 18]
```diff
@@ -0,0 +1,170 @@
...
+
+namespace WasmEdge {
+namespace Loader {
+
+Expect<void> Loader::loadInstantiateArg(
+    AST::InstantiateArg<AST::SortIndex<AST::Sort>> &Arg) {
+  // syntax `(with n si)`
+  //
+  // instantiatearg ::= n:<string>  si:<sortidx>
+  if (auto Res = FMgr.readName(); !Res) {
```
Comment (q82419) [inline lib/loader/ast/instance.cpp:15]: Reverse the if-else statement.

File: lib/loader/ast/instance.cpp [hunk 19]
```diff
@@ -0,0 +1,170 @@
...
+    return Unexpect(Res);
+  } else {
+    Arg.getName() = *Res;
+  }
+  return loadSortIndex(Arg.getIndex());
+}
+
+Expect<void> Loader::loadInlineExport(AST::InlineExport<AST::Sort> &Exp) {
+  if (auto Res = FMgr.readName(); !Res) {
```
Comment (q82419) [inline lib/loader/ast/instance.cpp:24]: Reverse the if-else statement.

File: include/ast/component/sort.h [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+
+#include "ast/expression.h"
+#include "ast/type.h"
+
+#include <vector>
+
+namespace WasmEdge {
+namespace AST {
+
+enum class CoreSort {
```
Comment (q82419) [inline include/ast/component/sort.h:24]: Assign the values.

File: include/ast/component/sort.h [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+  Func,
+  Table,
+  Memory,
+  Global,
+  Type,
+  Module,
+  Instance,
+};
+enum class SortCase {
```
Comment (q82419) [inline include/ast/component/sort.h:33]: Assign the values.

File: lib/loader/ast/sort.cpp [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+  //        | 0x03                  => type
+  //        | 0x04                  => component
+  //        | 0x05                  => instance
+  auto Tag = FMgr.readByte();
+  if (!Tag) {
+    spdlog::error(ErrInfo::InfoAST(ASTNodeAttr::Sort));
+    return Unexpect(Tag);
+  }
+  switch (*Tag) {
+  case 0x00: {
```
Comment (q82419) [inline lib/loader/ast/sort.cpp:23]: Use the assigned values in the enum class.

**Review by q82419 — CHANGES_REQUESTED** [review]
1. Add the `Component` in `AST` namespace.
2. Rebase onto the master branch.

**Review by q82419 — CHANGES_REQUESTED** [review]
1. Add the component model proposal enum.
2. Add the proposal description in tools.
3. Add the warning message (experimental) when turn-on and detect a component model unit loaded.

**Review by q82419 — APPROVED** [review]

**Comment by dannypsnl** [comment]
@hydai This is ready to merge.

**Review by q82419 — APPROVED** [review]

**Pull request merged**